pg_search/src/postgres/storage/linked_items.rs

```diff
@@ -346,6 +357,158 @@ impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> LinkedItemList<
+                    let new_offsetno = new_page.append_item(item, size, 0);
+                    assert!(new_offsetno != pg_sys::InvalidOffsetNumber);
+                }
+                offsetno += 1;
+            }
+
+            *new_page.special_mut::<BM25PageSpecialData>() =
+                (*page.special::<BM25PageSpecialData>()).clone();
+            blockno = page.next_blockno();
+            previous_new_buffer = Some(new_buffer);
+        }
+
+        AtomicGuard {
+            original: self,
+            original_header_lock,
+            cloned: Some(cloned),
+        }
+    }
+}
+
+pub struct AtomicGuard<'s, T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> {
+    original: &'s mut LinkedItemList<T>,
+    original_header_lock: BufferMut,
+    cloned: Option<LinkedItemList<T>>,
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> Deref for AtomicGuard<'_, T> {
+    type Target = LinkedItemList<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.cloned.as_ref().expect("Cannot deref after commit!")
+    }
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> DerefMut for AtomicGuard<'_, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.cloned
+            .as_mut()
+            .expect("Cannot deref_mut after commit!")
+    }
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> AtomicGuard<'_, T> {
+    pub fn commit(&mut self) {
+        let mut cloned = self.cloned.take().expect("Cannot commit twice!");
+        // Update our header page to point to the new start block, and return the old one.
+        let mut blockno = {
+            // Open both header pages.
+            let mut original_header_page = self.original_header_lock.page_mut();
+            let cloned_header_buffer = cloned.bman.get_buffer(cloned.header_blockno);
+            let cloned_header_page = cloned_header_buffer.page();
+
+            // Capture our old start page, then overwrite our metadata.
+            let original_metadata = original_header_page.contents_mut::<LinkedListData>();
+            let old_start_blockno = original_metadata.start_blockno;
+            *original_metadata = cloned_header_page.contents::<LinkedListData>();
+
+            // Finally, garbage collect the cloned header block.
+            cloned.bman.return_to_fsm(cloned_header_buffer);
+
+            old_start_blockno
+        };
+
+        // And then collect our old contents, which are no longer reachable.
+        while blockno != pg_sys::InvalidBlockNumber {
+            let buffer = self.original.bman().get_buffer(blockno);
+            blockno = buffer.page().next_blockno();
+            self.original.bman.return_to_fsm(buffer);
+        }
+    }
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> Drop for AtomicGuard<'_, T> {
+    fn drop(&mut self) {
+        let Some(mut cloned) = self.cloned.take() else {
```

Comment:
Because of how pgrx does stack unwinding, it'd be possible for `drop` to be called when we're no longer in a transaction state.  A raised ERROR/panic elsewhere in the callstack could cause this.

So this function needs an early return again to:

```rust
unsafe {
    if !pg_sys::IsTransactionState() { return; }
}
```

I realize that means we'd leak these non-committed pages forever, but surely that's better than crashing, which is what would happen if we tried to do any buffer-related work while no longer being in a transaction.

Reply:
And I reckon `drop` here wouldn't play nice with my suggestion above to have `.commit()` consume `self`.  But I think we could account for that by adding a `did_commit: bool` flag to `AtomicGuard` and returning early here if it's true.

Reply:
> So this function needs an early return to:

Will do.

> But I think we could account for that by adding a `did_commit: bool` flag to `AtomicGuard` and returning early here if it's true.

I'm not understanding this suggestion I think: `self.cloned().take()` is already acting as our early-return in this case. Are you saying that there would be a way for `commit` to take `self` and still have `Drop`?